be/src/runtime/http_result_writer.cpp

```diff
@@ -0,0 +1,173 @@
+    _init_profile();
+    if (nullptr == _sinker) {
+        return Status::InternalError("sinker is NULL pointer.");
+    }
+
+    return Status::OK();
+}
+
+void HttpResultWriter::_init_profile() {
+    _append_chunk_timer = ADD_TIMER(_parent_profile, "AppendChunkTime");
+    _convert_tuple_timer = ADD_CHILD_TIMER(_parent_profile, "TupleConvertTime", "AppendChunkTime");
+    _result_send_timer = ADD_CHILD_TIMER(_parent_profile, "ResultRendTime", "AppendChunkTime");
+    _sent_rows_counter = ADD_COUNTER(_parent_profile, "NumSentRows", TUnit::UNIT);
+}
+
+Status HttpResultWriter::append_chunk(Chunk* chunk) {
+    return Status::NotSupported("HttpResultWriter doesn't support non-pipeline engine");
+}
+
+Status HttpResultWriter::close() {
+    COUNTER_SET(_sent_rows_counter, _written_rows);
+    return Status::OK();
+}
+
+StatusOr<TFetchDataResultPtrs> HttpResultWriter::process_chunk(Chunk* chunk) {
+    SCOPED_TIMER(_append_chunk_timer);
+    int num_rows = chunk->num_rows();
+    std::vector<TFetchDataResultPtr> results;
+
+    Columns result_columns;
+    // Step 1: compute expr
+    int num_columns = _output_expr_ctxs.size();
+    result_columns.reserve(num_columns);
+
+    for (int i = 0; i < num_columns; ++i) {
+        ASSIGN_OR_RETURN(ColumnPtr column, _output_expr_ctxs[i]->evaluate(chunk));
+        column = _output_expr_ctxs[i]->root()->type().type == TYPE_TIME
+                         ? ColumnHelper::convert_time_column_from_double_to_str(column)
+                         : column;
+        result_columns.emplace_back(std::move(column));
+    }
+
+    // Step 2: convert chunk to http json row format row by row
+    {
+        TRY_CATCH_ALLOC_SCOPE_START()
+        row_str.reserve(128);
+        size_t current_bytes = 0;
+        int current_rows = 0;
+        SCOPED_TIMER(_convert_tuple_timer);
+        auto result = std::make_unique<TFetchDataResult>();
+        auto& result_rows = result->result_batch.rows;
+        result_rows.resize(num_rows);
+
+        for (int i = 0; i < num_rows; ++i) {
+            row_str.append("{\"data\":[");
+            for (auto& result_column : result_columns) {
+                std::string row = cast_type_to_json_str(result_column, i).value();
+                row_str.append(row);
+                if (result_column != result_columns[num_columns - 1]) {
+                    row_str.append(",");
+                }
+            }
+            row_str.append("]}\n");
+
+            size_t len = row_str.size();
+
+            if (UNLIKELY(current_bytes + len >= _max_row_buffer_size)) {
+                result_rows.resize(current_rows);
+                results.emplace_back(std::move(result));
+
+                result = std::make_unique<TFetchDataResult>();
+                result_rows = result->result_batch.rows;
+                result_rows.resize(num_rows - i);
+
+                current_bytes = 0;
+                current_rows = 0;
+            }
+
+            // VLOG_ROW << "written row:" << row_str;
```

Comment:
suggest to pre-allocate a large enough row_str for re-use, when the row_str is cooked well, just copy it into result_rows[current_rows]

Reply:
change to: result_rows[current_rows] = std::move(row_str);

Reply:
change to: result_rows[current_rows] = std::move(row_str);